File: mobile/ios/ShiftPlanner/ShiftPlanner/Features/Manager/Employees/Repositories/APIEmployeeManagementRepository.swift
import Foundation

enum EmployeeManagementRepositoryError: LocalizedError {
    case missingCompany
    case employeeRemovalUnavailable

    var errorDescription: String? {
        switch self {
        case .missingCompany:
            return localized("Company context is missing.", "Не найден контекст компании.")
        case .employeeRemovalUnavailable:
            return localized("The backend does not support removing employees from the company yet.", "Бэкенд пока не поддерживает удаление сотрудников из компании.")
        }
    }
}

private struct EmployeeManagementEmployeeResponseDTO: Decodable {
    let id: Int
    let publicId: String
    let fullName: String
    let email: String
    let role: UserRole
    let branchId: Int?
    let branch: EmployeeManagementBranchSummaryDTO?
    let positionId: Int?
    let positionTitle: String

    enum CodingKeys: String, CodingKey {
        case id
        case publicId = "public_id"
        case fullName = "full_name"
        case email
        case role
        case branchId = "branch_id"
        case branch
        case positionId = "position_id"
        case positionTitle = "position_title"
    }
}

private struct EmployeeManagementBranchSummaryDTO: Decodable {
    let id: Int
    let name: String
}

private struct EmployeeManagementPositionResponseDTO: Decodable {
    let id: Int
    let title: String
    let companyId: Int?

    enum CodingKeys: String, CodingKey {
        case id
        case title
        case companyId = "company_id"
    }
}

private struct EmployeeManagementBranchResponseDTO: Decodable {
    let id: Int
    let name: String
    let companyId: Int

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case companyId = "company_id"
    }
}

private struct EmployeeManagementPositionCreateRequestDTO: Encodable {
    let title: String
    let companyId: Int

    enum CodingKeys: String, CodingKey {
        case title
        case companyId = "company_id"
    }
}

private struct EmployeeManagementPositionUpdateRequestDTO: Encodable {
    let positionId: Int?

    enum CodingKeys: String, CodingKey {
        case positionId = "position_id"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(positionId, forKey: .positionId)
    }
}

private struct EmployeeManagementBranchUpdateRequestDTO: Encodable {
    let branchId: Int?

    enum CodingKeys: String, CodingKey {
        case branchId = "branch_id"
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(branchId, forKey: .branchId)
    }
}

private struct LinkUserRequestDTO: Encodable {
    let userPublicId: String
    let branchId: Int?
    let positionId: Int?

    enum CodingKeys: String, CodingKey {
        case userPublicId = "user_public_id"
        case branchId = "branch_id"
        case positionId = "position_id"
    }
}

private struct LinkedEmployeeResponseDTO: Decodable {
    let id: Int
    let publicId: String
    let fullName: String
    let email: String
    let branchId: Int?
    let positionId: Int?

    enum CodingKeys: String, CodingKey {
        case id
        case publicId = "public_id"
        case fullName = "full_name"
        case email
        case branchId = "branch_id"
        case positionId = "position_id"
    }

    func asManagedEmployee() -> ManagedEmployee {
        ManagedEmployee(
            id: id,
            publicId: publicId,
            fullName: fullName,
            email: email,
            role: .employee,
            branchId: branchId,
            branchName: nil,
            positionId: positionId,
            positionTitle: nil
        )
    }
}

private struct EmployeeWorkLimitsDTO: Codable {
    let maxHoursPerWeek: Int
    let maxHoursPerDay: Int

    enum CodingKeys: String, CodingKey {
        case maxHoursPerWeek = "max_hours_per_week"
        case maxHoursPerDay = "max_hours_per_day"
    }
}

private struct ManagerRequestDTO: Decodable {
    let id: Int
    let publicId: String
    let fullName: String
    let email: String
    let managerRole: String
    let membershipStatus: String

    enum CodingKeys: String, CodingKey {
        case id
        case publicId = "public_id"
        case fullName = "full_name"
        case email
        case managerRole = "manager_role"
        case membershipStatus = "membership_status"
    }
}

private struct EmployeeRequestDTO: Decodable {
    let id: Int
    let publicId: String
    let fullName: String
    let email: String
    let branchId: Int?
    let positionId: Int?
    let isActive: Bool

    enum CodingKeys: String, CodingKey {
        case id
        case publicId = "public_id"
        case fullName = "full_name"
        case email
        case branchId = "branch_id"
        case positionId = "position_id"
        case isActive = "is_active"
    }
}

extension EmployeeManagementEmployeeResponseDTO {
    func asManagedEmployee() -> ManagedEmployee {
        ManagedEmployee(
            id: id,
            publicId: publicId,
            fullName: fullName,
            email: email,
            role: role,
            branchId: branchId,
            branchName: branch?.name,
            positionId: positionId,
            positionTitle: positionTitle.isEmpty ? nil : positionTitle
        )
    }
}

extension EmployeeManagementPositionResponseDTO {
    func asManagedPosition() -> ManagedPosition {
        ManagedPosition(id: id, title: title)
    }
}

extension EmployeeManagementBranchResponseDTO {
    func asManagedBranch() -> ManagedBranch {
        ManagedBranch(id: id, name: name)
    }
}

final class APIEmployeeManagementRepository: EmployeeManagementRepository {
    let capabilities = EmployeeManagementCapabilities(
        canCreatePosition: true,
        canAssignPosition: true,
        canRemovePosition: true,
        canRemoveEmployee: false
    )

    private let apiClient: APIClient
    private let companyId: Int?

    init(companyId: Int?, apiClient: APIClient = .shared) {
        self.companyId = companyId
        self.apiClient = apiClient
    }

    func fetchEmployees() async throws -> [ManagedEmployee] {
        let request = apiClient.makeRequest(
            path: "employees/",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [EmployeeManagementEmployeeResponseDTO].self)
        return response.map { $0.asManagedEmployee() }
    }

    func fetchBranches() async throws -> [ManagedBranch] {
        let request = apiClient.makeRequest(
            path: "companies/branches",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [EmployeeManagementBranchResponseDTO].self)
        return response.map { $0.asManagedBranch() }
    }

    func fetchPositions() async throws -> [ManagedPosition] {
        guard let companyId else {
            throw EmployeeManagementRepositoryError.missingCompany
        }

        let request = apiClient.makeRequest(
            path: "positions/",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [EmployeeManagementPositionResponseDTO].self)

        return response
            .filter { $0.companyId == companyId }
            .map { $0.asManagedPosition() }
            .sorted { $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending }
    }

    func addPosition(title: String, currentPositions: [ManagedPosition]) async throws -> [ManagedPosition] {
        guard let companyId else {
            throw EmployeeManagementRepositoryError.missingCompany
        }

        let body = try JSONEncoder().encode(
            EmployeeManagementPositionCreateRequestDTO(
                title: title,
                companyId: companyId
            )
        )
        let request = apiClient.makeRequest(
            path: "positions/",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let created = try await apiClient.send(request, as: EmployeeManagementPositionResponseDTO.self).asManagedPosition()

        var updated = currentPositions
        updated.append(created)
        return updated.sorted { $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending }
    }

    func removePosition(
        _ position: ManagedPosition,
        from employees: [ManagedEmployee],
        positions: [ManagedPosition]
    ) async throws -> EmployeeManagementSnapshot {
        let request = apiClient.makeRequest(
            path: "positions/\(position.id)",
            method: "DELETE",
            requiresAuthorization: true
        )
        try await apiClient.sendWithoutResponseBody(request)

        return EmployeeManagementSnapshot(
            employees: employees,
            positions: positions.filter { $0.id != position.id }
        )
    }

    func assignBranch(
        _ branchId: Int?,
        to employee: ManagedEmployee,
        in employees: [ManagedEmployee]
    ) async throws -> [ManagedEmployee] {
        let body = try JSONEncoder().encode(
            EmployeeManagementBranchUpdateRequestDTO(branchId: branchId)
        )
        let request = apiClient.makeRequest(
            path: "employees/\(employee.id)/branch",
            method: "PATCH",
            body: body,
            requiresAuthorization: true
        )
        let updatedEmployee = try await apiClient.send(request, as: EmployeeManagementEmployeeResponseDTO.self).asManagedEmployee()

        return employees.map { existingEmployee in
            guard existingEmployee.id == employee.id else { return existingEmployee }
            return updatedEmployee
        }
    }

    func assignPosition(
        _ positionId: Int?,
        to employee: ManagedEmployee,
        in employees: [ManagedEmployee]
    ) async throws -> [ManagedEmployee] {
        let body = try JSONEncoder().encode(
            EmployeeManagementPositionUpdateRequestDTO(positionId: positionId)
        )
        let request = apiClient.makeRequest(
            path: "employees/\(employee.id)/position",
            method: "PATCH",
            body: body,
            requiresAuthorization: true
        )
        let updatedEmployee = try await apiClient.send(request, as: EmployeeManagementEmployeeResponseDTO.self).asManagedEmployee()

        return employees.map { existingEmployee in
            guard existingEmployee.id == employee.id else { return existingEmployee }
            return updatedEmployee
        }
    }

    func removeEmployee(_ employee: ManagedEmployee, from employees: [ManagedEmployee]) async throws -> [ManagedEmployee] {
        employees.filter { $0.id != employee.id }
    }

    // MARK: - Linking

    func linkEmployeeByPublicId(publicId: String, branchId: Int?, positionId: Int?) async throws -> ManagedEmployee {
        let body = try JSONEncoder().encode(
            LinkUserRequestDTO(userPublicId: publicId, branchId: branchId, positionId: positionId)
        )
        let request = apiClient.makeRequest(
            path: "companies/me/link-user",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let linked = try await apiClient.send(request, as: LinkedEmployeeResponseDTO.self)
        return linked.asManagedEmployee()
    }

    // MARK: - Work limits

    func fetchWorkLimits(employeeId: Int) async throws -> WorkLimits {
        let request = apiClient.makeRequest(
            path: "employees/\(employeeId)/work-limits",
            method: "GET",
            requiresAuthorization: true
        )
        let dto = try await apiClient.send(request, as: EmployeeWorkLimitsDTO.self)
        return WorkLimits(maxHoursPerWeek: dto.maxHoursPerWeek, maxHoursPerDay: dto.maxHoursPerDay)
    }

    func updateWorkLimits(employeeId: Int, maxHoursPerWeek: Int, maxHoursPerDay: Int) async throws -> WorkLimits {
        let body = try JSONEncoder().encode(
            EmployeeWorkLimitsDTO(maxHoursPerWeek: maxHoursPerWeek, maxHoursPerDay: maxHoursPerDay)
        )
        let request = apiClient.makeRequest(
            path: "employees/\(employeeId)/work-limits",
            method: "PATCH",
            body: body,
            requiresAuthorization: true
        )
        let dto = try await apiClient.send(request, as: EmployeeWorkLimitsDTO.self)
        return WorkLimits(maxHoursPerWeek: dto.maxHoursPerWeek, maxHoursPerDay: dto.maxHoursPerDay)
    }

    // MARK: - Join requests

    func fetchManagerRequests() async throws -> [PendingManagerRequest] {
        let request = apiClient.makeRequest(
            path: "companies/me/manager-requests",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [ManagerRequestDTO].self)
        return response
            .filter { $0.membershipStatus == "pending" }
            .map {
                PendingManagerRequest(id: $0.id, fullName: $0.fullName, email: $0.email, managerRole: $0.managerRole)
            }
    }

    func acceptManagerRequest(id: Int) async throws {
        try await postWithoutResponse(path: "companies/me/manager-requests/\(id)/accept")
    }

    func declineManagerRequest(id: Int) async throws {
        try await postWithoutResponse(path: "companies/me/manager-requests/\(id)/decline")
    }

    func fetchEmployeeRequests() async throws -> [PendingEmployeeRequest] {
        let request = apiClient.makeRequest(
            path: "companies/me/employee-requests",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [EmployeeRequestDTO].self)
        return response
            .filter { !$0.isActive }
            .map {
                PendingEmployeeRequest(
                    id: $0.id, fullName: $0.fullName, email: $0.email,
                    positionId: $0.positionId, branchId: $0.branchId
                )
            }
    }

    func acceptEmployeeRequest(id: Int) async throws {
        try await postWithoutResponse(path: "companies/me/employee-requests/\(id)/accept")
    }

    func declineEmployeeRequest(id: Int) async throws {
        try await postWithoutResponse(path: "companies/me/employee-requests/\(id)/decline")
    }

    private func postWithoutResponse(path: String) async throws {
        let request = apiClient.makeRequest(
            path: path,
            method: "POST",
            requiresAuthorization: true
        )
        try await apiClient.sendWithoutResponseBody(request)
    }
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/employees/ApiEmployeeManagementRepository.kt
package com.froggyriia.shiftplanner.data.employees

import com.froggyriia.shiftplanner.data.network.ApiClient
import com.froggyriia.shiftplanner.data.network.EmployeeBranchUpdateRequestDto
import com.froggyriia.shiftplanner.data.network.EmployeeCreateRequestDto
import com.froggyriia.shiftplanner.data.network.EmployeePositionUpdateRequestDto
import com.froggyriia.shiftplanner.data.network.LinkUserRequestDto
import com.froggyriia.shiftplanner.data.network.EmployeeWorkLimitsDto
import com.froggyriia.shiftplanner.data.network.EmployeeBranchesUpdateDto
import com.froggyriia.shiftplanner.data.network.CompanyUserPublicIdRequestDto
import com.froggyriia.shiftplanner.data.network.PositionCreateRequestDto
import com.froggyriia.shiftplanner.domain.model.ManagedBranch
import com.froggyriia.shiftplanner.domain.model.ManagedBranchAssignment
import com.froggyriia.shiftplanner.domain.model.ManagedEmployee
import com.froggyriia.shiftplanner.domain.model.ManagedEmployeeCalendar
import com.froggyriia.shiftplanner.domain.model.ManagedEmployeeCalendarShift
import com.froggyriia.shiftplanner.domain.model.ManagedPosition
import com.froggyriia.shiftplanner.domain.model.PendingEmployeeRequest
import com.froggyriia.shiftplanner.domain.model.PendingManagerRequest
import com.froggyriia.shiftplanner.domain.model.WorkLimits

class ApiEmployeeManagementRepository(
    private val apiClient: ApiClient,
    private val companyId: Int?
) : EmployeeManagementRepository {

    override suspend fun fetchEmployees(): List<ManagedEmployee> = wrap {
        apiClient.api.getEmployees().map { it.toDomain() }
            .sortedBy { it.fullName }
    }

    override suspend fun fetchBranches(): List<ManagedBranch> = wrap {
        apiClient.api.getBranches().map { ManagedBranch(id = it.id, name = it.name) }
    }

    override suspend fun fetchPositions(): List<ManagedPosition> = wrap {
        apiClient.api.getPositions()
            .filter { it.companyId == companyId }
            .map { it.toDomain() }
            .sortedBy { it.title }
    }

    override suspend fun createEmployee(
        fullName: String,
        email: String,
        positionId: Int,
        branchId: Int?
    ): ManagedEmployee = wrap {
        val created = apiClient.api.createEmployee(
            EmployeeCreateRequestDto(fullName = fullName, email = email, positionId = positionId)
        ).toDomain()

        if (branchId != null) {
            apiClient.api.updateEmployeeBranch(
                id = created.id,
                request = EmployeeBranchUpdateRequestDto(branchId = branchId)
            ).toDomain()
        } else {
            created
        }
    }

    override suspend fun deleteEmployee(employeeId: Int) = wrap {
        apiClient.api.deleteEmployee(employeeId)
        Unit
    }

    override suspend fun assignPosition(employeeId: Int, positionId: Int?): ManagedEmployee = wrap {
        apiClient.api.updateEmployeePosition(
            id = employeeId,
            request = EmployeePositionUpdateRequestDto(positionId = positionId)
        ).toDomain()
    }

    override suspend fun assignBranch(employeeId: Int, branchId: Int?): ManagedEmployee = wrap {
        apiClient.api.updateEmployeeBranch(
            id = employeeId,
            request = EmployeeBranchUpdateRequestDto(branchId = branchId)
        ).toDomain()
    }

    override suspend fun createPosition(title: String, companyId: Int): ManagedPosition = wrap {
        apiClient.api.createPosition(
            PositionCreateRequestDto(title = title, companyId = companyId)
        ).toDomain()
    }

    override suspend fun deletePosition(positionId: Int) = wrap {
        apiClient.api.deletePosition(positionId)
        Unit
    }

    override suspend fun linkEmployeeByPublicId(
        publicId: String,
        branchId: Int?,
        positionId: Int?
    ): ManagedEmployee = wrap {
        apiClient.api.linkUserToCompany(
            LinkUserRequestDto(userPublicId = publicId, branchId = branchId, positionId = positionId)
        ).toDomain()
    }

    override suspend fun fetchWorkLimits(employeeId: Int): WorkLimits = wrap {
        apiClient.api.getEmployeeWorkLimits(employeeId).let {
            WorkLimits(maxHoursPerWeek = it.maxHoursPerWeek, maxHoursPerDay = it.maxHoursPerDay)
        }
    }

    override suspend fun updateWorkLimits(
        employeeId: Int,
        maxHoursPerWeek: Int,
        maxHoursPerDay: Int
    ): WorkLimits = wrap {
        apiClient.api.updateEmployeeWorkLimits(
            id = employeeId,
            request = EmployeeWorkLimitsDto(
                maxHoursPerWeek = maxHoursPerWeek,
                maxHoursPerDay = maxHoursPerDay
            )
        ).let { WorkLimits(maxHoursPerWeek = it.maxHoursPerWeek, maxHoursPerDay = it.maxHoursPerDay) }
    }

    // ── Pending join requests ─────────────────────────────────────────────────

    override suspend fun fetchManagerRequests(): List<PendingManagerRequest> = wrap {
        apiClient.api.getManagerRequests()
            .filter { it.membershipStatus == "pending" }
            .map { it.toDomain() }
    }

    override suspend fun acceptManagerRequest(id: Int) = wrap {
        apiClient.api.acceptManagerRequest(id); Unit
    }

    override suspend fun declineManagerRequest(id: Int) = wrap {
        apiClient.api.declineManagerRequest(id); Unit
    }

    override suspend fun fetchEmployeeRequests(): List<PendingEmployeeRequest> = wrap {
        apiClient.api.getEmployeeRequests()
            .filter { !it.isActive }
            .map { it.toDomain() }
    }

    override suspend fun acceptEmployeeRequest(id: Int) = wrap {
        apiClient.api.acceptEmployeeRequest(id); Unit
    }

    override suspend fun declineEmployeeRequest(id: Int) = wrap {
        apiClient.api.declineEmployeeRequest(id); Unit
    }

    override suspend fun addManagerByPublicId(publicId: String): PendingManagerRequest = wrap {
        apiClient.api.addManagerByPublicId(CompanyUserPublicIdRequestDto(userPublicId = publicId)).toDomain()
    }

    override suspend fun fetchEmployeeBranches(employeeId: Int): List<ManagedBranchAssignment> = wrap {
        apiClient.api.getEmployeeBranches(employeeId).map {
            ManagedBranchAssignment(id = it.id, name = it.name, isPrimary = it.isPrimary)
        }
    }

    override suspend fun replaceEmployeeBranches(
        employeeId: Int,
        branchIds: List<Int>,
        primaryBranchId: Int
    ): List<ManagedBranchAssignment> = wrap {
        apiClient.api.replaceEmployeeBranches(
            employeeId,
            EmployeeBranchesUpdateDto(branchIds = branchIds, primaryBranchId = primaryBranchId)
        ).map { ManagedBranchAssignment(id = it.id, name = it.name, isPrimary = it.isPrimary) }
    }

    override suspend fun fetchEmployeeCalendar(employeeId: Int): ManagedEmployeeCalendar = wrap {
        val dto = apiClient.api.getEmployeeCalendarSummary(employeeId)
        val shifts = dto.shifts.map { s ->
            ManagedEmployeeCalendarShift(
                date = s.date,
                startMinutes = timeToMinutes(s.startTime),
                endMinutes = timeToMinutes(s.endTime),
                status = s.status ?: ""
            )
        }.sortedWith(compareBy({ it.date }, { it.startMinutes }))
        ManagedEmployeeCalendar(
            employeeName = dto.employee.fullName,
            totalShifts = dto.workload?.totalShifts ?: shifts.size,
            totalHours = dto.workload?.totalHours ?: 0.0,
            shifts = shifts
        )
    }

    private fun timeToMinutes(time: String): Int {
        val parts = time.split(":").mapNotNull { it.toIntOrNull() }
        return if (parts.size >= 2) parts[0] * 60 + parts[1] else 0
    }

    private suspend fun <T> wrap(block: suspend () -> T): T {
        return try {
            block()
        } catch (error: Throwable) {
            throw Exception(apiClient.userMessage(error))
        }
    }
}
